W-11862525-first-draft-for-beta-site-build-CP (#114)
mulesoft/docs-apikit: modules/ROOT/pages/apikit-odatav4-extension-reference.adoc

= APIkit for OData v4 Module Reference
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Learn the fundamental building blocks about dependency, namespace, and schema information in APIkit for OData v4 to collaborate, build applications, and maintain codebases.

Reviewer: ```suggestion
Learn about the fundamental building blocks, dependencies, namespaces, and schema information in APIkit for OData v4 to collaborate, build applications, and maintain codebases.
```
Please check if this suggestion keeps the intended meaning. The original sentence was a bit odd.

Author: In the original sentence, the building blocks are (1) dependency, (2) namespace, and (3) schema information.
In your version, there are (1) building blocks, (2) dependency, (3) namespace, and (4) schema information.
Actually, the article has a dependency title, a namespace title, and it provides schema information.
We can reword it to:
"Learn the fundamentals about dependency, namespace, and schema information in APIkit for OData v4 to collaborate, build applications, and maintain codebases".
Would it be a better option?
Thanks!

----

== Configurations

=== Config

Default configuration

==== Parameters

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Name | Type | Description | Default Value | Required
| Name | String | Name for this configuration. Connectors reference the configuration with this name | | *x*
| Api Definition | String | API definition |  | *x*
| Name | String | Identifier of this element used to reference it in other components |  | *x*
|===

==== Associated Operations

* xref:apikit-odatav4-extension-reference#route[]
* xref:apikit-odatav4-extension-reference#serialize-entity[]
* xref:apikit-odatav4-extension-reference#serialize-entity-collection[]
* xref:apikit-odatav4-extension-reference#transform-to-sql-select[]

==== Associated Sources

* xref:apikit-odatav4-extension-reference#on-entity-collection-request[RequestEntityCollectionListener]
* xref:apikit-odatav4-extension-reference#on-entity-request[RequestEntityListener]

== Operations

== Route

`<apikit-odata:route>`

=== Parameters

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use | | *x*

> | Content | Any |  |  #[payload] | 

Reviewer: Global comment: for consistency, all code values and functions/operations like `#[payload]`, `WHERE`, `$skip`, `$search`, `true`, `false`, etc. should be in backticks.

Author: OK! Thanks!

`<apikit-odata:transform-to-sql-select>`

=== Parameters

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use | | *x*
| Entity Set Name | String | Table name in the database for the requested entity set |  | *x*
| Database Schema | String | Database schema name to which the entity set belongs. If not empty, it is prefixed to the table name in the WHERE clause |  | 
| Page Size | Number |  |  | 
| Config Ref | ConfigurationProvider | Name of the configuration used to execute this component |  | *x*
| Streaming Strategy | 

* xref:apikit-odatav4-extension-reference#repeatable-in-memory-stream[RepeatableInMemoryStream]
* xref:apikit-odatav4-extension-reference#repeatable-file-store-stream[RepeatableFileStoreStream]
* non-repeatable-stream 

| Configure if repeatable streams and their behavior are used |  | 
| Entity Type Name | String | Entity Data Model specifies the entity type name |  | *x*
| Entity Type Fields | Array of String | List of property fields from the Entity type |  | *x*
| Entity Type Keys | Object | Map of entity key names and values |  | *x*

> | System Query Options | Object | System Query Options map. The grammar and syntax rules for system query options are defined in [OData-ABNF], to simplify usability the initial dollar sign is removed to avoid escaping withing DataWeave, in example to access System Query Option $skip user uses the following expression #[attributes.systemQueryOptions.'skip'] |  | *x*

Reviewer: ```suggestion
| System Query Options | Object | System Query Options map. The grammar and syntax rules for system query options are defined in [OData-ABNF]. To simplify usability, the initial dollar sign is removed to avoid escaping within DataWeave. For example, to access System Query Option, $skip uses the following expression #[attributes.systemQueryOptions.'skip'] |  | *x*
```
Not sure if the suggestion is correct, the sentence is long and the last part is confusing. Please check.

Author: Great catch! Thanks!

* APIKIT-ODATA:MALFORMED_URI 
* APIKIT-ODATA:NOT_IMPLEMENTED 
* APIKIT-ODATA:TRANSFORMATION 

== Sources

== On Entity Collection Request

`<apikit-odata:request-entity-collection-listener>`

=== Parameters

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use | | *x*
| Output MIME Type | String | MIME type of the payload that this operation outputs |  | 
| Output Encoding | String | Encoding of the payload that this operation outputs |  | 
| Config Ref | ConfigurationProvider | Name of the configuration used to execute this component |  | *x*

> | Primary Node Only | Boolean | If this source must only be executed on the primary node when running in Cluster |  | 

Reviewer: ```suggestion
| Primary Node Only | Boolean | Specify if this source executes only on the primary node when running in Cluster |  | 
```

Author: OK! Thanks!

|===

=== OData Http Attributes

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Field | Type | Description | Default Value | Required
| Headers | Object |  |  |
| Method | Enumeration, one of:

* GET
* POST
* PUT
* PATCH
* DELETE 

|  |  |
| Raw Base Uri | String |  |  |
| Raw Odata Path | String |  |  |
| Raw Query Path | String |  |  |
| Raw Request Uri | String |  |  |
| Scheme | Enumeration, one of:

* HTTP
* HTTPS 

|  |  |
|===

=== OData Request Attributes

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Field | Type | Description | Default Value | Required
| Custom Query Options | Object |  |  |
| Entity Type Fields | Array of String |  |  |
| Entity Type Keys | Object |  |  |
| Entity Type Name | String |  |  |
| System Query Options | Object |  |  |
|===

=== Repeatable in Memory Stream

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Field | Type | Description | Default Value | Required
| Initial Buffer Size | Number | Amount of memory allocated to consume the stream and provide random access to it. If the stream has more data than buffer capacity, it expands according to the bufferSizeIncrement attribute, with an upper limit of maxInMemorySize. |  |

> | Buffer Size Increment | Number | Expansion of the buffer size if exceeding its initial size. Setting a value of zero or lower means that the buffer does not expand, meaning that a STREAM_MAXIMUM_SIZE_EXCEEDED error is raised when the buffer gets full. |  |

Reviewer: ```suggestion
| Buffer Size Increment | Number | Expansion of the buffer size if exceeding its initial size. If you set a value of zero or lower, the buffer does not expand and raises a STREAM_MAXIMUM_SIZE_EXCEEDED error when the buffer gets full. |  |
```

Author: OK! Thanks!

> | Max Buffer Size | Number | This is the maximum amount of memory used. If more than that is used, then a STREAM_MAXIMUM_SIZE_EXCEEDED error is raised. A value lower or equal to zero means no limit. |  |

Reviewer: ```suggestion
| Max Buffer Size | Number | The maximum amount of memory for the buffer to use. When the memory usage exceeds this limit, a STREAM_MAXIMUM_SIZE_EXCEEDED error is raised. A value lower or equal to zero means no limit. |  |
```

Author: OK! Thanks!

| Buffer Unit | Enumeration, one of:

* BYTE
* KB
* MB
* GB 

| Unit to express these attributes |  |
|===

=== Repeatable File Store Stream

[cols="1a,2a,3a,4a,5a", options="header"]
|===
| Field | Type | Description | Default Value | Required

> | In Memory Size | Number | Maximum memory that the stream use to keep data in memory. If more is consumed, it starts to buffer the content on disk |  |

Reviewer: ```suggestion
| In Memory Size | Number | Maximum memory that the stream uses to keep data in memory. If more memory is consumed, the stream starts to buffer the content on the disk |  |
```

Author: OK! Thanks!
